In [932]:
import pandas as pd
import numpy as np

In [933]:
df = pd.read_csv("Titanic-Dataset.csv")

## 1. Initial Data Exploration

First, lets inspect the structure of the dataset

In [934]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


In [935]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [936]:
df.shape

(891, 12)

In [937]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='str')

In [938]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object

Now we know, that we have 12 columns, with types int, str and float. Also we know, that we have 891 passengers here. The dataset contains information about Titanic passengers, including demographic information, ticket class and survival status.

## 2. Counting Missing values

Lets check, how much and what values we dont have in this DataSet

In [939]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [940]:
df.isna().mean() * 100

PassengerId     0.000000
Survived        0.000000
Pclass          0.000000
Name            0.000000
Sex             0.000000
Age            19.865320
SibSp           0.000000
Parch           0.000000
Ticket          0.000000
Fare            0.000000
Cabin          77.104377
Embarked        0.224467
dtype: float64

Age and Cabin contain a significant number of missing values, while Embarked has only a small number of missing entries.

## 3. Base information, Duplicates, Unique values

In [941]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [942]:
df.describe(include='str')

,Name,Sex,Ticket,Cabin,Embarked
count,891,891,891,204,889
unique,891,2,681,147,3
top,"Braund, Mr. Owen Harris",male,347082,G6,S
freq,1,577,7,4,644


In [943]:
df.duplicated().sum()

np.int64(0)

In [944]:
df.nunique()

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             88
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin          147
Embarked         3
dtype: int64

The dataset contains no duplicated rows.

Age contains about 19.9% missing values, while Cabin has about
77.1% missing values. Embarked contains very few missing values.

Passenger ages range from 0.42 to 80 years, with a median age of 28.

Fare appears to have a wide and asymmetric distribution:
the median fare is about 14.45, while the maximum reaches 512.33.

Most passengers in the dataset are male, and S is the most common
value in the Embarked column.

## 4. Univariate Analysis

Now let's analyze them one characteristic at a time.

In [945]:
df['Survived'].value_counts()

Survived
0    549
1    342
Name: count, dtype: int64

In [946]:
df['Survived'].value_counts(normalize=True) * 100

Survived
0    61.616162
1    38.383838
Name: proportion, dtype: float64

From this, we conclude that only about 38% of the people survived

In [947]:
df['Sex'].value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

In [948]:
df['Sex'].value_counts(normalize=True)

Sex
male      0.647587
female    0.352413
Name: proportion, dtype: float64

In [949]:
df['Pclass'].value_counts()

Pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [950]:
df['Pclass'].value_counts(normalize=True)

Pclass
3    0.551066
1    0.242424
2    0.206510
Name: proportion, dtype: float64

More than half of the people were third-class passengers

In [951]:
df['Embarked'].value_counts()

Embarked
S    644
C    168
Q     77
Name: count, dtype: int64

In [952]:
df['Embarked'].value_counts(normalize=True)

Embarked
S    0.724409
C    0.188976
Q    0.086614
Name: proportion, dtype: float64

In [953]:
df[['Age', 'Fare']].describe()

,Age,Fare
count,714.000000,891.000000
mean,29.699118,32.204208
std,14.526497,49.693429
min,0.420000,0.000000
25%,20.125000,7.910400
50%,28.000000,14.454200
75%,38.000000,31.000000
max,80.000000,512.329200


75% of passengers have a fare of ≤ 31

but max = 512.33

Moreover:

mean = 32.20
median = 14.45

The mean is much higher than the median.

## 5. Survival Analysis

In [954]:
df.groupby('Sex')['Survived'].mean() * 100

Sex
female    74.203822
male      18.890815
Name: Survived, dtype: float64

In [955]:
df.groupby('Pclass')['Survived'].mean() * 100

Pclass
1    62.962963
2    47.282609
3    24.236253
Name: Survived, dtype: float64

In [956]:
df.groupby(['Sex','Pclass'])['Survived'].mean() * 100

Sex     Pclass
female  1         96.808511
        2         92.105263
        3         50.000000
male    1         36.885246
        2         15.740741
        3         13.544669
Name: Survived, dtype: float64

In [957]:
df.groupby(['Survived','Sex'])['Age'].median()

Survived  Sex   
0         female    24.5
          male      29.0
1         female    28.0
          male      28.0
Name: Age, dtype: float64

In [958]:
df.groupby('Survived')['Age'].median()

Survived
0    28.0
1    28.0
Name: Age, dtype: float64

In [959]:
df['FamilySize'] = df['SibSp'] + df['Parch'] + 1
df["FamilySize"].value_counts().sort_index()

FamilySize
1     537
2     161
3     102
4      29
5      15
6      22
7      12
8       6
11      7
Name: count, dtype: int64

In [960]:
df.groupby('FamilySize')['Survived'].mean() * 100

FamilySize
1     30.353818
2     55.279503
3     57.843137
4     72.413793
5     20.000000
6     13.636364
7     33.333333
8      0.000000
11     0.000000
Name: Survived, dtype: float64

In [961]:
df["FamilySize"].value_counts().sort_index()

FamilySize
1     537
2     161
3     102
4      29
5      15
6      22
7      12
8       6
11      7
Name: count, dtype: int64

### Conclusions

1. Women had a substantially higher survival rate than men.

2. Survival rate was highest among first-class passengers and
   lowest among third-class passengers.

3. Female passengers in first and second class had similarly high
   survival rates. Male passengers had considerably lower survival
   rates across all classes.

4. The median age was 28 years for both survivors and non-survivors,
   so median age alone does not show a clear difference between them.

5. Most passengers traveled without family members. FamilySize = 4
   had the highest observed survival rate, although some family-size
   groups contain relatively few passengers, so these percentages
   should be interpreted carefully.